### PR TITLE
[Mailer] Added Sendinblue bridge

### PR DESCRIPTION
**Pull request by drixs6o9**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT
| Doc PR        | [https://github.com/symfony/symfony-docs/pull/14272](https://github.com/symfony/symfony-docs/pull/14272)
| Recipe | [https://github.com/symfony/recipes/pull/822](https://github.com/symfony/recipes/pull/822)

N.B. I had a little help from [Pierre TONDEREAU](https://github.com/ptondereau) for the API.

### CODE REVIEW

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueTransportFactory.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        if (!\in_array($dsn->getScheme(), $this->getSupportedSchemes(), true)) {
+            throw new UnsupportedSchemeException($dsn, 'sendinblue', $this->getSupportedSchemes());
+        }
+
+        switch ($dsn->getScheme()) {
+            default:
+            case 'sendinblue':
+            case 'sendinblue+smtp':
+                $transport = sprintf(self::NAMESPACE, 'SendinblueSmtpTransport');
```
[the remark this answers is not in the document]
Comment (drixs6o9) [reply src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueTransportFactory.php:39]: Indeed, I'll push this modification.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueApiTransport.php [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @author Yann LUCAS
+ */
+final class SendinblueApiTransport extends AbstractApiTransport
+{
+    private const SENDINBLUE_API_HOST = 'api.sendinblue.com';
```
[the remark this answers is not in the document]
Comment (drixs6o9) [reply src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueApiTransport.php:33]: I have followed Sendgrid exemple.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.3.0
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Sendinblue/CHANGELOG.md:4]: Could be 5.2 as I think we can add it in 5.2 in time.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/composer.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            "email": "fabien@symfony.com"
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=7.2.5",
+        "symfony/mailer": "^4.4|^5.0"
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Sendinblue/composer.json:20]: Should be ^5.1 (as you are using MetadataHeader which is not part of earlier versions)

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpTransport.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    public function __construct(string $username, string $password, ?EventDispatcherInterface $dispatcher = null, ?LoggerInterface $logger = null)
+    {
+        parent::__construct(
+            self::SENDINBLUE_SMTP_HOST,
+            self::SENDINBLUE_SMTP_PORT,
+            self::SENDINBLUE_TLS_ENABLED,
+            $dispatcher,
+            $logger
+        );
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpTransport.php:35]: Can be on one line

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpsTransport.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+namespace Symfony\Component\Mailer\Bridge\Sendinblue\Transport;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Mailer\Transport\Smtp\EsmtpTransport;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * @author Yann LUCAS
+ */
+final class SendinblueSmtpsTransport extends EsmtpTransport
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpsTransport.php:21]: This class should not be needed. The EsmtpTransport is able to determine the port automatically.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueTransportFactory.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                    ->setPort($dsn->getPort())
+                ;
+        }
+
+        return new $transport(
+            $this->getUser($dsn),
+            $this->getPassword($dsn),
+            $this->dispatcher,
+            $this->logger
+        );
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueTransportFactory.php:54]: should be on one line

**Review by fabpot — CHANGES_REQUESTED** [review]
You also need to register the transport in `src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer_transports.xml`, and add it to `src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php`, `src/Symfony/Component/Mailer/Exception/UnsupportedSchemeException.php`, and  `src/Symfony/Component/Mailer/Transport.php` as well.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/CHANGELOG.md (hunk 3 above)
Comment (drixs6o9) [reply src/Symfony/Component/Mailer/Bridge/Sendinblue/CHANGELOG.md]: I have changed this also for symfony-docs and recipe.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/composer.json (hunk 4 above)
Comment (drixs6o9) [reply src/Symfony/Component/Mailer/Bridge/Sendinblue/composer.json]: Done.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpTransport.php (hunk 5 above)
Comment (drixs6o9) [reply src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpTransport.php]: Done.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpsTransport.php (hunk 6 above)
Comment (drixs6o9) [reply src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpsTransport.php]: I have removed this class.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueTransportFactory.php (hunk 7 above)
Comment (drixs6o9) [reply src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueTransportFactory.php]: Done.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/README.md [hunk 8]
```diff
@@ -0,0 +1,12 @@
+Sendinblue Bridge
+=================
+
+Provides Sendinblue integration for Symfony Mailer.
+
```
Comment (noniagriconomie) [inline src/Symfony/Component/Mailer/Bridge/Sendinblue/README.md:5]: can you add like this, https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Mailer/Bridge/Mailjet/README.md,
a minimal config example? i think it is helpfull :)

Comment (fabpot) [reply src/Symfony/Component/Mailer/Bridge/Sendinblue/README.md:5]: Having the format of the DSN is indeed a good addition.

Comment (drixs6o9) [reply src/Symfony/Component/Mailer/Bridge/Sendinblue/README.md:5]: It's done. I've also added custom header example which is possible with Sendinblue.

File: src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpTransport.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Mailer\Transport\Smtp\EsmtpTransport;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * @author Yann LUCAS
+ */
+final class SendinblueSmtpTransport extends EsmtpTransport
+{
+    const SENDINBLUE_SMTP_HOST = 'smtp-relay.sendinblue.com';
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpTransport.php:23]: These 3 consts can be inlined IMHO or at least be marked private.

Comment (drixs6o9) [reply src/Symfony/Component/Mailer/Bridge/Sendinblue/Transport/SendinblueSmtpTransport.php]: Yes i changed for inline values. There is no other usages for those constants.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @drixs6o9.

**Comment by drixs6o9** [comment]
> Thank you @drixs6o9.

Thank you too. And also @nicolas-grekas and @noniagriconomie

**Comment by noniagriconomie** [comment]
@drixs6o9 👍 i saw that this provider do also have an [SMS endpoint](https://developers.sendinblue.com/reference#sendtransacsms)

maybe it could be added (if PR can be reviewed/merged like this one) :)

**Comment by drixs6o9** [comment]
> @drixs6o9  i saw that this provider do also have an [SMS endpoint](https://developers.sendinblue.com/reference#sendtransacsms)
> 
> maybe it could be added (if PR can be reviewed/merged like this one) :)

Yes, it's a good idea. I'll see this soon.

**Pull request merged**